src/dtree.ml

```diff
@@ -0,0 +1,499 @@
+
+  (** [get_cons l] extracts, sorts and uniqify terms that are tree
+      constructors in [l].  The actual tree constructor (of type
+      {!type:treecons}) is returned along the original term. *)
+  let get_cons : term list -> (treecons * term) list = fun telst ->
+    let keep_treecons e =
+      if is_treecons e then Some(treecons_of_term e, e) else None in
+    let tc_fst_cmp (tca, _) (tcb, _) = tc_compare tca tcb in
+    telst |> List.filter_map keep_treecons |> List.sort_uniq tc_fst_cmp
+
+  (** [contains_abst l] returns whether list of terms [l] contains an
+      abstraction. *)
+  let rec contains_abst : term list -> bool = function
+    | [] -> false
+    | Abst(_, _) :: _  -> true
+    | _          :: xs -> contains_abst xs
+
+  (** [contains_var l] returns whether list of terms [l] contains a pattern
+      variable with a slot assigned. *)
+  let rec contains_var : term list -> bool = function
+    | []                        -> false
+    | Patt(Some(_), _, _) :: _  -> true
+    | _                   :: xs -> contains_var xs
+
+  (** [in_rhs] returns whether a list of term contains a term needed in the
+      right hand side of a rule. *)
+  let rec in_rhs : term list -> bool = function
+    | []                       -> false
+    | Patt(Some(_), _, _) :: _ -> true
+    | _ :: xs                  -> in_rhs xs
+
+  (** [enrich_env_builder c v r] adds an entry into the environment builder of
+      [r] based on the nature of term at column [c] and with [v] variables met
+      so far. *)
+  let enrich_env_builder : int -> int -> rule -> rule = fun ci var_met r ->
+    let t, _ = r.lhs.(ci) in
+    match fst (get_args t) with
+    | Patt(Some(i), _, _) ->
+        { r with env_builder = (var_met, i) :: r.env_builder}
+    | _                   -> r
+
+(** [spec_filter p e] returns whether a line been inspected on element [e]
+    (from a pattern matrix) must be kept when specializing the matrix on
+    pattern [p].
+    @raise Invalid_argument when the element [e] is ill formed. *)
+  let spec_filter : term -> term -> bool = fun pat hd ->
+    let h, args = get_args hd in
+    let ph, pargs = get_args pat in
+    match ph, h with
+    | Symb(_, _), Symb(_, _)
+    | Vari(_)   , Vari(_)       -> List.same_length args pargs && eq ph h
+    | _         , Patt(_, _, e) ->
+        if args <> [] then invalid_arg "ClauseMat.spec_filter" else
+        let b = Bindlib.bind_mvar (Array.map to_tvar e) (lift ph) in
+        Bindlib.is_closed b
+    | _         , _             -> false
+
+  (** [spec_transform p e] transform element [e] (from a lhs) when
+      specializing against pattern [p]. *)
+  let spec_transform : term -> component -> component array =
+    fun pat (t, p) ->
+      let h, args = Basics.get_args t in
+      match h with
+      | Symb(_, _)
+      | Vari(_)       ->
+          let np = Subterm.sub p in
+          args |> List.to_seq |> Subterm.tag ~empty:np |> Array.of_seq
```

Comment:
Is Seq useful here?

Reply:
Please add a comment (in each file) on the first occurrence of |> as I'm not sure whether this notation is well known.